xlive/Blam/Engine/input/input_xinput.cpp

```diff
@@ -1,66 +1,217 @@
+	// Originally 20
+	// We set to 4 and check if if the controller is plugged in
+	XINPUT_VIBRATION vibration{};
+
+	for (uint32 i = 0; i < 4; i++)
+	{
+		input_device* device = g_xinput_devices[i];
+		if (device && controller_button_state_get((e_controller_index)i)->plugged_in)
+		{
+			device->XSetState(&vibration);
+		}
+	}
+	return;
 }
 
 void input_xinput_update_rumble_state(void)
 {
-    bool suppress_rumble = g_input_feedback_suppress || *input_suppress_global_get();
-    if (!game_in_progress() || game_time_get_paused())
-    {
-        suppress_rumble = true;
-    }
-
-    if (controller_profile_get(_controller_index_0)->field_1658 != 1)
-    {
-        suppress_rumble = true;
-    }
-
-    if (controller_button_state_get(_controller_index_0)->plugged_in)
-    {
-        g_xinput_vibration.wLeftMotorSpeed = (suppress_rumble ? 0 : g_vibration_state[_controller_index_0].left);
-        g_xinput_vibration.wRightMotorSpeed = (suppress_rumble ? 0 : g_vibration_state[_controller_index_0].right);
-        xinput_device* device = g_xinput_devices[*g_main_controller_index];
-        
-        if (device)
-        {
-            device->set_state(&g_xinput_vibration);
-        }
-    }
-
-    return;
+	bool suppress_rumble = g_input_feedback_suppress || *input_suppress_global_get();
+	if (!game_in_progress() || game_time_get_paused())
+	{
+		suppress_rumble = true;
+	}
+
+	if (controller_profile_get(_controller_index_0)->field_1658 != 1)
+	{
+		suppress_rumble = true;
+	}
+
+	if (controller_button_state_get(_controller_index_0)->plugged_in)
+	{
+		g_xinput_vibration.wLeftMotorSpeed = (suppress_rumble ? 0 : g_vibration_state[_controller_index_0].wLeftMotorSpeed);
+		g_xinput_vibration.wRightMotorSpeed = (suppress_rumble ? 0 : g_vibration_state[_controller_index_0].wRightMotorSpeed);
+		input_device* device = g_xinput_devices[*g_main_controller_index];
+
+		if (device)
+		{
+			device->XSetState(&g_xinput_vibration);
+		}
+	}
+
+	return;
+}
+
+
+void input_xinput_update_button(uint8* frames_down, uint16* msec_down, bool button_down, uint16 duration_ms)
+{
+	if (button_down)
+	{
+		*frames_down = MIN(*frames_down + 1, UINT8_MAX);
+		*msec_down = MIN(*msec_down + duration_ms, UINT16_MAX);
+	}
+	else
+	{
+		*frames_down = 0;
+		*msec_down = 0;
+	}
+}
+
+void input_xinput_update_trigger(uint8* last_msec_down, bool trigger_down, uint8 current_msec_down)
+{
+	if (trigger_down)
+	{
+		uint8 msec_down = CLAMP_LOWER(current_msec_down, 0, 32);
+		*last_msec_down = MAX(msec_down, *last_msec_down);
+	}
+	else
+	{
+		uint8 msec_down = CLAMP_UPPER(current_msec_down, 64, 255);
+		*last_msec_down = MIN(msec_down, *last_msec_down);
+	}
+}
+
+int16 input_xinput_adjust_thumb_axis_deadzone(int16 thumb_axis, int16 thumb_deadzone)
+{
+	if (thumb_axis > thumb_deadzone)
+		return  0x7FFF * (thumb_axis - thumb_deadzone) / (0x7FFF - thumb_deadzone);
+
+	if (thumb_axis < -thumb_deadzone)
+		return ~0x7FFF * (thumb_axis + thumb_deadzone) / (~0x7FFF + thumb_deadzone);
```

Comment:
~0x7FFF (NOT 0x7FFF) is not -32767, but -32768. Should be -0x7FFF instead, unless it is intentional.